### ISSUE TITLE
Unable to Register a Custom Dataset!! 

### ISSUE THREAD

**Issue by bansilol** [facebookresearch/detectron2#4362]
Hi Guys, I was trying to register the BDD Dataset as a builtin dataset for Detectron2, but unfortunately the list of reistered datasets gets appended while running the script but later when calling thw Dataset it says the Dataset is not registered.
# Instructions To Reproduce the Issue:

### a. I  added a new file to the dectron2/data/datasets folder named bdd100k.py, this script is supposed to get the dicts for the dataset. The contents of this file are as follows.

`
def get_bdd_dicts(img_dir, json_dir, dataset_type):
    json_file = os.path.join(json_dir)

    with open(json_file) as f:
        image_counter = 0
        imgs_anns = json.load(f)

    dataset_dicts = []
    for item in imgs_anns:
        record = {}
        record["file_name"] = item['name']
        image_filename = os.path.join(img_dir + dataset_type, item['name'])
        #print(image_filename)  # Should be something like '/raid/Datasets/Radspot/free_space_detection/bdd100k/images/100k/train/0000f77c-6257be58.jpg'
        height, width = cv2.imread(image_filename).shape[:2]

        record["file_name"] = image_filename
        record["height"] = height
        record["width"] = width
        record["image_id"] = image_counter

        image_counter += 1
        annotations = []
        key='labels'
        if key in item:
            for n in range(len(item['labels'])):
                bbox=item['labels'][n]['box2d']
                annotation = {
                            "bbox": list(bbox.values()),
                            "bbox_mode": BoxMode.XYXY_ABS,
                            "category_id": BDD100K_OD_CATEGORIES[item['labels'][n]['category']],
                            "iscrowd": 0
                            }
                annotations.append(annotation)
                record["annotations"] = annotations
        else:
            annotation = {
                            "bbox": None,
                            "bbox_mode": BoxMode.XYXY_ABS,
                            "iscrowd": 0
                            }
            annotations.append(annotation)
            record["annotations"] = annotations 
        dataset_dicts.append(record)    
        
        #print(dataset_dicts)
        # [{'file_name': '/home/ubuntu/FinalProject/bdd-data/bdd100k/images/100k/train/0000f77c-6257be58.jpg', 'height': 720, 'width': 1280, 'image_id': 0, 'annotations': [{'bbox': [1125.902264, 133.184488, 1156.978645, 210.875445], 'bbox_mode': <BoxMode.XYXY_ABS: 0>, 'category_id': 6, 'iscrowd': 0}]}]
        # Output exactly matches Detectron2 Tutorial.ipynb on Google colab
        
    return dataset_dicts`

### b. I added a function to register this dataset in the builtin.py
`  
   def register_bdd100k(root):
       img_dir = os.path.join(root, "images/100k/")
       json_dir= os.path.join(root, "labels/bdd100k_det_20_labels_trainval/bdd100k/labels/det_20/")
       remove_dataset_name = ["bdd100k_det_20_train","bdd100k_det_20_val"]
       for name in remove_dataset_name:
            if name in DatasetCatalog.list():
               DatasetCatalog.remove(name)
               MetadataCatalog.remove(name)
               print("dataset de-registered")
        for d in ["train", "val"]:
            name = "bdd100k_det_20_" + d
            json_dir = os.path.join(json_dir,"det_"+ d +".json")
            DatasetCatalog.register(
                name, lambda x=img_dir, y=json_dir: get_bdd_dicts(x, y, d)
             )
            MetadataCatalog.get(name).set(thing_classes=BDD100K_OD_CATEGORIES)
    
    if __name__.endswith("builtin"):
       _root= os.path.expanduser(os.getenv("DETECTRON2_DATASETS", "datasets"))
       register_bdd100k(_root)`

## 2. What exact command you run: 
To register the dataset, I run the file __init__.py which runs the builtin.py script and registers my dataset using the `register_bdd100k(root)` function.
Then to check if the dataset is registered I use the training code from the colab notebook provided by you, there I get the error that the dataset is not registered. 

## 3. __Full logs__ or other relevant observations:
```
<Exception has occurred: KeyError
"Dataset 'bdd100k_det_20_train' is not registered! Available datasets are: coco_2014_train, coco_2014_val, coco_2014_minival, coco_2014_valminusminival, coco_2017_train, coco_2017_val, coco_2017_test, coco_2017_test-dev, coco_2017_val_100, keypoints_coco_2014_train, keypoints_coco_2014_val, keypoints_coco_2014_minival, keypoints_coco_2014_valminusminival, keypoints_coco_2017_train, keypoints_coco_2017_val, keypoints_coco_2017_val_100, coco_2017_train_panoptic_separated, coco_2017_train_panoptic_stuffonly, coco_2017_train_panoptic, coco_2017_val_panoptic_separated, coco_2017_val_panoptic_stuffonly, coco_2017_val_panoptic, coco_2017_val_100_panoptic_separated, coco_2017_val_100_panoptic_stuffonly, coco_2017_val_100_panoptic, lvis_v1_train, lvis_v1_val, lvis_v1_test_dev, lvis_v1_test_challenge, lvis_v0.5_train, lvis_v0.5_val, lvis_v0.5_val_rand_100, lvis_v0.5_test, lvis_v0.5_train_cocofied, lvis_v0.5_val_cocofied, cityscapes_fine_instance_seg_train, cityscapes_fine_sem_seg_train, cityscapes_fine_instance_seg_val, cityscapes_fine_sem_seg_val, cityscapes_fine_instance_seg_test, cityscapes_fine_sem_seg_test, cityscapes_fine_panoptic_train, cityscapes_fine_panoptic_val, voc_2007_trainval, voc_2007_train, voc_2007_val, voc_2007_test, voc_2012_trainval, voc_2012_train, voc_2012_val, ade20k_sem_seg_train, ade20k_sem_seg_val"

The above exception was the direct cause of the following exception:

  File "/home/bansi/detectron2/demo/demo_bdd100k_train.py", line 26, in <module>
    trainer = DefaultTrainer(cfg)>
```

## Expected behavior:

As per the tutorial in the colab notebook the dataset should have gotten registered and when I try to train a network and call that dataset detectron2 should have been able to find it. Can you please help me with this issue? I wish to run a custom training so how do I get my dataset permanently in the DatasetCatalog? As you can see in the logs above even after doing the registeration I get the error that the dataset is not registered.
